src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/MultiBlockSystem.java

```diff
@@ -0,0 +1,57 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+package org.terasology.ligthandshadow.componentsystem.controllers;
+
+import java.util.Collection;
+
+import org.joml.Vector3i;
+import org.terasology.engine.entitySystem.entity.EntityRef;
+import org.terasology.engine.entitySystem.event.EventPriority;
+import org.terasology.engine.entitySystem.event.ReceiveEvent;
+import org.terasology.engine.entitySystem.prefab.Prefab;
+import org.terasology.engine.entitySystem.systems.BaseComponentSystem;
+import org.terasology.engine.entitySystem.systems.RegisterMode;
+import org.terasology.engine.entitySystem.systems.RegisterSystem;
+import org.terasology.engine.logic.characters.events.AttackEvent;
+import org.terasology.engine.logic.health.EngineDamageTypes;
+import org.terasology.engine.logic.inventory.ItemComponent;
+import org.terasology.engine.world.BlockEntityRegistry;
+import org.terasology.module.health.events.DoDamageEvent;
+import org.terasology.engine.registry.In;
+import org.terasology.multiBlock2.component.MultiBlockMainComponent;
+import org.terasology.multiBlock2.component.MultiBlockMemberComponent;
+
+/**
+ * Damages all the blocks in a multiBlock Structure when any block in the structure is damaged.
+ */
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class MultiBlockSystem extends BaseComponentSystem {
+    @In
+    private BlockEntityRegistry blockEntityRegistry;
+
+    @ReceiveEvent(priority = EventPriority.PRIORITY_CRITICAL)
+    public void onDamageDone(AttackEvent event, EntityRef entity) {
+        Collection<Vector3i> blocks = null;
+        int damage = 1;
+        Prefab damageType = EngineDamageTypes.PHYSICAL.get();
+        ItemComponent item = event.getDirectCause().getComponent(ItemComponent.class);
+        if (item != null) {
+            damage = item.baseDamage;
+            if (item.damageType != null) {
+                damageType = item.damageType;
+            }
+        }
+        if (entity.hasComponent(MultiBlockMainComponent.class)) {
+            blocks = entity.getComponent(MultiBlockMainComponent.class).getMultiBlockMembers();
+        } else if (entity.hasComponent(MultiBlockMemberComponent.class)) {
+            blocks = blockEntityRegistry.getBlockEntityAt(entity.getComponent(MultiBlockMemberComponent.class).getMainBlockLocation()).getComponent(MultiBlockMainComponent.class).getMultiBlockMembers();
+        }
```

Comment:
I'd do this in two separate steps:

1. find the main multi-block, if any
2. derive the `blocks` from that main block

To make the code easier to comprehend I'd additionally move this to a method which returns an `Optional<Collection<Vector3i>>` with the multi-block member positions, if any multi-block is hit.

As it does not make any difference for us whether we iterate over an empty collection (and thus don't damage any other blocks) or check if the collection is present we may even get rid of the optional here and just return a collection that may or may not contain any entries.